cpp/src/arrow/compute/util_internal.h

```diff
@@ -27,5 +29,56 @@ void CheckAlignment(const void* ptr) {
   ARROW_DCHECK(reinterpret_cast<uint64_t>(ptr) % sizeof(T) == 0);
 }
 
+/// Storage used to allocate temporary vectors of a batch size.
+/// Temporary vectors should resemble allocating temporary variables on the stack
+/// but in the context of vectorized processing where we need to store a vector of
+/// temporaries instead of a single value.
+class ARROW_EXPORT TempVectorStack {
+  template <typename>
+  friend class TempVectorHolder;
+
+ public:
+  Status Init(MemoryPool* pool, int64_t size);
+
+  int64_t AllocatedSize() const { return top_; }
```

Comment:
This is not related to the purpose of this PR but rather something remaining in https://github.com/apache/arrow/pull/41352#discussion_r1598969656 and too trivial for another PR.